### ISSUE TITLE
Stack overflow?

### ISSUE THREAD

**Issue by klauer** [pcdshub/lucid#67]
Not the website:

```
[2020-06-04 15:09:40,088] [ERROR   ] - An uncaught exception happened: wrapped C/C++ object of type Connection has been deleted
Traceback (most recent call last):
  File "_ctypes/callbacks.c", line 234, in 'calling callback function'
  File "/reg/g/pcds/pyps/apps/dev/pythonpath/epics/ca.py", line 720, in _onConnectionEvent
    timestamp=time.time())
  File "/reg/g/pcds/pyps/apps/dev/pythonpath/epics/ca.py", line 263, in run_connection_callbacks
    callback(pvname=self.pvname, chid=chid_int, conn=self.conn)
  File "/reg/g/pcds/pyps/apps/dev/pythonpath/epics/pv.py", line 48, in wrapped
    return func(self, *args, **kwargs)
  File "/reg/g/pcds/pyps/apps/dev/pythonpath/epics/pv.py", line 325, in __on_connect
    conn_cb(pvname=self.pvname, conn=conn, pv=self)
  File "/reg/g/pcds/pyps/apps/dev/pythonpath/pydm/data_plugins/epics_plugins/pyepics_plugin_component.py", line 111, in send_connection_state
    self.connection_state_signal.emit(conn)
RuntimeError: wrapped C/C++ object of type Connection has been deleted
[2020-06-04 15:09:40,107] [ERROR   ] - An uncaught exception happened: wrapped C/C++ object of type Connection has been deleted
Traceback (most recent call last):
  File "_ctypes/callbacks.c", line 234, in 'calling callback function'
  File "/reg/g/pcds/pyps/apps/dev/pythonpath/epics/ca.py", line 769, in _onAccessRightsEvent
    entry.run_access_event_callbacks(read, write)
  File "/reg/g/pcds/pyps/apps/dev/pythonpath/epics/ca.py", line 236, in run_access_event_callbacks
    callback(ra, wa)
  File "/reg/g/pcds/pyps/apps/dev/pythonpath/epics/pv.py", line 48, in wrapped
    return func(self, *args, **kwargs)
  File "/reg/g/pcds/pyps/apps/dev/pythonpath/epics/pv.py", line 290, in __on_access_rights_event
    cb(read_access, write_access, pv=self)
  File "/reg/g/pcds/pyps/apps/dev/pythonpath/pydm/data_plugins/epics_plugins/pyepics_plugin_component.py", line 102, in send_access_state
    self.write_access_signal.emit(write_access)
RuntimeError: wrapped C/C++ object of type Connection has been deleted
Fatal Python error: Cannot recover from stack overflow.
```

I may have closed the window before this happened, but I don't recall (it was during a meeting).

```
Current thread 0x00007f0eba2d9740 (most recent call first):
  File "/reg/g/pcds/pyps/conda/py36/envs/dev/lib/python3.6/linecache.py", line 95 in updatecache
  File "/reg/g/pcds/pyps/conda/py36/envs/dev/lib/python3.6/linecache.py", line 47 in getlines
  File "/reg/g/pcds/pyps/conda/py36/envs/dev/lib/python3.6/linecache.py", line 16 in getline
  File "/reg/g/pcds/pyps/conda/py36/envs/dev/lib/python3.6/traceback.py", line 286 in line
  File "/reg/g/pcds/pyps/conda/py36/envs/dev/lib/python3.6/traceback.py", line 364 in extract
  File "/reg/g/pcds/pyps/conda/py36/envs/dev/lib/python3.6/traceback.py", line 509 in __init__
  File "/reg/g/pcds/pyps/conda/py36/envs/dev/lib/python3.6/traceback.py", line 104 in print_exception
  File "/reg/g/pcds/pyps/conda/py36/envs/dev/lib/python3.6/logging/__init__.py", line 535 in formatException
  File "/reg/g/pcds/pyps/conda/py36/envs/dev/lib/python3.6/logging/__init__.py", line 585 in format
  File "/reg/g/pcds/pyps/conda/py36/envs/dev/lib/python3.6/logging/__init__.py", line 840 in format
  File "/reg/g/pcds/pyps/conda/py36/envs/dev/lib/python3.6/logging/__init__.py", line 994 in emit
  File "/reg/g/pcds/pyps/conda/py36/envs/dev/lib/python3.6/logging/__init__.py", line 865 in handle
  File "/reg/g/pcds/pyps/conda/py36/envs/dev/lib/python3.6/logging/__init__.py", line 1516 in callHandlers
  File "/reg/g/pcds/pyps/conda/py36/envs/dev/lib/python3.6/logging/__init__.py", line 1454 in handle
  File "/reg/g/pcds/pyps/conda/py36/envs/dev/lib/python3.6/logging/__init__.py", line 1444 in _log
  File "/reg/g/pcds/pyps/conda/py36/envs/dev/lib/python3.6/logging/__init__.py", line 1337 in error
  File "/reg/g/pcds/pyps/conda/py36/envs/dev/lib/python3.6/logging/__init__.py", line 1343 in exception
  File "/reg/g/pcds/pyps/apps/dev/lucid/lucid/main_window.py", line 195 in handle_error
  File "/reg/g/pcds/pyps/apps/dev/pythonpath/pydm/exception.py", line 182 in raise_to_operator
  File "/reg/g/pcds/pyps/apps/dev/pythonpath/pydm/utilities/__init__.py", line 320 in wrapper
  File "/reg/g/pcds/pyps/apps/dev/lucid/lucid/main_window.py", line 196 in handle_error
  File "/reg/g/pcds/pyps/apps/dev/pythonpath/pydm/exception.py", line 182 in raise_to_operator
  File "/reg/g/pcds/pyps/apps/dev/pythonpath/pydm/utilities/__init__.py", line 320 in wrapper
  File "/reg/g/pcds/pyps/apps/dev/lucid/lucid/main_window.py", line 196 in handle_error
  File "/reg/g/pcds/pyps/apps/dev/pythonpath/pydm/exception.py", line 182 in raise_to_operator
  File "/reg/g/pcds/pyps/apps/dev/pythonpath/pydm/utilities/__init__.py", line 320 in wrapper
  File "/reg/g/pcds/pyps/apps/dev/lucid/lucid/main_window.py", line 196 in handle_error
  File "/reg/g/pcds/pyps/apps/dev/pythonpath/pydm/exception.py", line 182 in raise_to_operator
  File "/reg/g/pcds/pyps/apps/dev/pythonpath/pydm/utilities/__init__.py", line 320 in wrapper
  File "/reg/g/pcds/pyps/apps/dev/lucid/lucid/main_window.py", line 196 in handle_error
 ```

Mostly full log on NFS `~klauer/lucid_failure.txt`

**Comment by hhslepicka** [comment]
Looking at the log it seems like we need to remove our callbacks before the shutdown/closing of the Connection. In this case the `Connection` object was destroyed and got a new info from pyepics on the callback which unleashed the mayhem...

On StackOverflow it seems like it was caused by the unhandled exception being handled by the `handle_error` which also caused an exception and all hell broke loose... 😢

**Comment by hhslepicka** [comment]
Part of this issue was addressed along with (#73). Now what remains is why the handle_error and raise_to_operator caused this stack overflow.

**Comment by hhslepicka** [comment]
I was able to reproduce this error locally... 
Here is how I did it:
- Using PyDM 1.10.1 (to skip the fix for the connection and access callback) and PyEPICS (3.4.2, not master which includes the fix for callbacks as well)
- Modify the launcher or any other part of LUCID to create successfully about 500 connections or more of PyDMChannel with successful connection callbacks.
- Once the connections are made, disconnect the channels
- Kill the IOC
- Restart the IOC...

This will cause a flood of `RuntimeError: wrapped C/C++ object of type Connection has been deleted` which were fixed with PyDM 1.10.2.

Now I need to find out a way to stop this from ever happening in the future.
